consensus/common/src/main/java/org/hyperledger/besu/consensus/common/validator/blockbased/BlockValidatorProvider.java

```diff
@@ -0,0 +1,84 @@
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.hyperledger.besu.consensus.common.validator.blockbased;
+
+import org.hyperledger.besu.consensus.common.BftValidatorOverrides;
+import org.hyperledger.besu.consensus.common.BlockInterface;
+import org.hyperledger.besu.consensus.common.EpochManager;
+import org.hyperledger.besu.consensus.common.validator.ValidatorProvider;
+import org.hyperledger.besu.consensus.common.validator.VoteProvider;
+import org.hyperledger.besu.ethereum.chain.Blockchain;
+import org.hyperledger.besu.ethereum.core.Address;
+import org.hyperledger.besu.ethereum.core.BlockHeader;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+public class BlockValidatorProvider implements ValidatorProvider {
+
+  private final VoteTallyCache voteTallyCache;
+  private final VoteProvider voteProvider;
+
+  public static ValidatorProvider forkingValidatorProvider(
+      final Blockchain blockchain,
+      final EpochManager epochManager,
+      final BlockInterface blockInterface,
+      final Map<Long, List<Address>> validatorForkMap) {
+    return new BlockValidatorProvider(blockchain, epochManager, blockInterface, validatorForkMap);
+  }
+
+  public static ValidatorProvider nonForkingValidatorProvider(
+      final Blockchain blockchain,
+      final EpochManager epochManager,
+      final BlockInterface blockInterface) {
+    return new BlockValidatorProvider(
+        blockchain, epochManager, blockInterface, Collections.emptyMap());
+  }
+
+  private BlockValidatorProvider(
+      final Blockchain blockchain,
+      final EpochManager epochManager,
+      final BlockInterface blockInterface,
+      final Map<Long, List<Address>> validatorForkMap) {
+    final VoteTallyUpdater voteTallyUpdater = new VoteTallyUpdater(epochManager, blockInterface);
+    final VoteProposer voteProposer = new VoteProposer();
+    this.voteTallyCache =
+        new ForkingVoteTallyCache(
+            blockchain,
+            voteTallyUpdater,
+            epochManager,
+            blockInterface,
+            new BftValidatorOverrides(validatorForkMap));
+    this.voteProvider = new BlockVoteProvider(voteTallyCache, voteProposer);
+  }
+
+  @Override
+  public Collection<Address> getValidatorsAtHead() {
+    return voteTallyCache.getVoteTallyAtHead().getValidators();
+  }
+
+  @Override
+  public Collection<Address> getValidatorsAfterBlock(final BlockHeader header) {
+    return voteTallyCache.getVoteTallyAfterBlock(header).getValidators();
+  }
+
+  @Override
+  public Optional<VoteProvider> getVoteProvider() {
+    return Optional.of(voteProvider);
```

Comment:
`Optional.ofNullable` ??

Reply:
VoteProvider is mandatory for BlockValidatorProvider and we create vote provider on line 67 so it will always exist